Add Gulpjs as a cloud build option

amaxwell01 opened the pull request:
* Added Gulpjs - http://gulpjs.com/
* Added gulp version command for a test ```gulp --version```
* Added gulp help command for a test ```gulp --help```
* Created a simple example gulp build file to test ```gulp build```

Philmod commented on gulpjs/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+# Use the base App Engine Docker image, based on debian jessie.
+FROM debian:8
```
Why not using the npm builder as the base?

Philmod commented on gulpjs/cloudbuild.yaml:
```diff
@@ -0,0 +1,32 @@
+# Gulp build
+- name: 'gcr.io/$PROJECT_ID/gulpjs'
+  args: ['build']
+  dir: 'gulpjs/examples/build'
```
That would be nice to have an example that produces a static file, and in a next step pushes it to a CDN or to GCS.
I believe that's a realistic use case of gulpjs?

Philmod commented on gulpjs/examples/build/package.json:
```diff
@@ -0,0 +1,14 @@
+  "description": "",
+  "main": "gulpfile.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Either you remove this script, or you add a npm test as a additional build step (but then don't `exit 1`).

Philmod commented on gulpjs/examples/build/package.json:
```diff
@@ -0,0 +1,14 @@
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
```
Please use the Apache License. https://github.com/GoogleCloudPlatform/cloud-builders/blob/master/LICENSE

skelterjohn commented on gulpjs/cloudbuild.yaml:
```diff
@@ -0,0 +1,32 @@
+# Step 0
+- name: 'gcr.io/cloud-builders/docker'
+  args: ['build', '--tag=gcr.io/$PROJECT_ID/gulpjs', '.']
+  dir: 'gulpjs'
```
this file is meant to be executed from this directory, not the top of the repository (omit "gulpjs" from the `dir` fields).

JustinBeckwith commented:
I think @GoogleCloudPlatform/node-team should take a look!

JustinBeckwith commented on gulpjs/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+# Install the latest LTS release of nodejs
+RUN mkdir /nodejs && curl https://nodejs.org/dist/v6.9.1/node-v6.9.1-linux-x64.tar.gz | tar xvzf - -C /nodejs --strip-components=1
+ENV PATH $PATH:/nodejs/bin
+RUN npm install -g gulp-cli
```
@jmdobry @ofrobots 

Typically people will install gulp as a dev dependency in their package.json.  If you install dev dependencies, it also symlinks the entrypoint so you can just invoke `gulp` in an npm script.  This is important for producing hermetic builds, and also so you can pin to a specific version of gulp that's not head.  We should never `npm install -g` a dev dependency like this for a user.  

On other words - I don't think this is the right way to make this builder at all.  A proper builder for node.js can be built with a single multi-purpose pipeline. 

Proposed Steps
- Run `npm install` (notice that I didn't add --production).  
- Run some user defined build step, possibly by convention:  `npm run build`
- Build the container
- Entrypoint is set to `npm start`

If we're feeling super fancy we could do a two container process where we trim the npm dependencies down to only the --prod ones in the actual runtime container. 

@ofrobots - this really intersects with the builder work y'all are already doing.  We should sync.

amaxwell-gpsw replied:
Agreed that maybe pinning it to head is not the best, we could switch to specific versions. 

Right now there are multiple ways in the community to trigger gulp builds within a project:
1. "make" routing to the node_module dependency command
2. npm script
3. running the script via a node file or command line

The goal that I was trying to make was to abstract the different use cases for developers so that they can simplify their build steps. For my teams we usually use #1 but that requires both a dev dependency install and a prod install(two steps). Abstracting the gulp from the build avoids the use of installing dev dependencies if you wanted to do something simple using gulp.

amaxwell-gpsw replied:
I'd love to hop on a call or chat if you guys are interested.

Philmod replied:
We are going to discuss it internally first. I'll come back to you.
If you could provide a real example, that would be very helpful.

amaxwell01 commented:
Addressed most of the feedback except for:

```
That would be nice to have an example that produces a static file, and in a next step pushes it to a CDN or to GCS.
I believe that's a realistic use case of gulpjs?
```

Which is something that I can explore tonight.

Philmod commented:
Hey Andrew,

I wrote an example on how to use Gulpjs with GCB: https://github.com/Philmod/google-container-gulp

Feel free to reach out for any question.

amaxwell-gpsw commented:
The code looks good @Philmod nicely done. Better example to use the package specific version of Gulp.
